### PR TITLE
kube-up.sh: drop unnecessary legacy mirror config, enable injecting registry mirror

### PR DESCRIPTION
**Pull request by BenTheElder**
#### What type of PR is this?

/kind failing-test

#### What this PR does / why we need it:

Enables using a pull-through cache for images in kube-up.sh

Part of resolving 5k node scale test issues https://github.com/kubernetes/kubernetes/issues/126366

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: cluster/gce/gci/configure-helper.sh [hunk 1]
```diff
@@ -3315,16 +3315,30 @@ oom_score = -999
...
-  endpoint = ["https://registry.k8s.io", "https://k8s.gcr.io",]
 [plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runc.options]
   SystemdCgroup = ${systemdCgroup}
+# enable hosts config
+[plugins."io.containerd.grpc.v1.cri".registry]
+  config_path = "/etc/containerd/certs.d"
 EOF
 
+  # DO NOT MERGE -- Testing
+  KUBERNETES_REGISTRY_PULL_THROUGH_HOST='https://us-west2-docker.pkg.dev/v2/k8s-artifacts-prod/images/'
```
Comment (BenTheElder) [inline cluster/gce/gci/configure-helper.sh:3326]: should probably instead test it with the planned private endpoint and tweak IAM to allow the other boskos projects ... but we can start with this test commit and public read.

NOTE: do not do this yourself, at some point we're going to lock down access to these backends :-)

File: cluster/gce/gci/configure-helper.sh [hunk 2]
```diff
@@ -3315,16 +3315,30 @@ oom_score = -999
...
-  endpoint = ["https://registry.k8s.io", "https://k8s.gcr.io",]
 [plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runc.options]
   SystemdCgroup = ${systemdCgroup}
+# enable hosts config
+[plugins."io.containerd.grpc.v1.cri".registry]
+  config_path = "/etc/containerd/certs.d"
 EOF
 
+  # DO NOT MERGE -- Testing
+  KUBERNETES_REGISTRY_PULL_THROUGH_HOST='https://us-central1-docker.pkg.dev/v2/k8s-infra-e2e-scale-5k-project/k8s-5k-scale-cache/'
```
Comment (BenTheElder) [inline cluster/gce/gci/configure-helper.sh:3326]: I wiped the images in this just before the last run but it doesn't look like we populated any, so something is not quite right here (or the timing was off, but I don't think so).

Need to investigate this further before proceeding.

File: cluster/gce/gci/configure-helper.sh [hunk 3]
```diff
@@ -3315,16 +3318,32 @@ oom_score = -999
...
+  # used for 5k node scale tests with private pull-through cache
+  set -x;
+  if [[ -n "${KUBERNETES_REGISTRY_PULL_THROUGH_HOST:-}" ]]; then
+    registry_config_dir="/etc/containerd/certs.d/registry.k8s.io"
+    mkdir -p "${registry_config_dir}"
+    {
+      # NOTE: we need literal double quotes around some of these values
+      echo 'server="'"${KUBERNETES_REGISTRY_PULL_THROUGH_HOST}"'"'
+      echo ''
+      echo '[hosts."'"${KUBERNETES_REGISTRY_PULL_THROUGH_HOST}"'"]'
```
Comment (BenTheElder) [inline cluster/gce/gci/configure-helper.sh:3340]: this should be host.

File: cluster/gce/gci/configure-helper.sh (hunk 2 above)
Comment (BenTheElder) [reply cluster/gce/gci/configure-helper.sh:3329]: after fixing https://github.com/kubernetes/kubernetes/pull/126448#discussion_r1696030091, I see many many images pulled.

**Comment by BenTheElder** [comment]
/test pull-kubernetes-e2e-gce
[This time the cache is set to only allow `allAuthenticatedUsers` so we'll see if the credential provider gets invoked or not, in theory]

**Comment by BenTheElder** [comment]
So ... credential provider does not work for this currently.

**Comment by BenTheElder** [comment]
Yeah, it's only trying to get an anonymous token https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/126448/pull-kubernetes-e2e-gce/1818064974405177344/artifacts/e2e-7222b98906-674b9-master/containerd.log

**Comment by BenTheElder** [comment]
I got auth working locally against the pull through mirror in a kind cluster. We'll have to get this plumbed through on the test-infra side.

File: cluster/gce/gci/configure-helper.sh [hunk 4]
```diff
@@ -3315,16 +3315,32 @@ oom_score = -999
   default_runtime_name = "runc"
 [plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runc]
   runtime_type = "io.containerd.runc.v2"
-[plugins."io.containerd.grpc.v1.cri".registry.mirrors."docker.io"]
```
Comment (BenTheElder) [inline cluster/gce/gci/configure-helper.sh:3318]: this format is deprecated and nothing should be pulling through k8s.gcr.io anymore anyhow

ditto dockerhub, our CI shouldn't be pulling from it, and GCR is deprecated

File: cluster/gce/gci/configure-helper.sh [hunk 5]
```diff
@@ -3315,16 +3315,32 @@ oom_score = -999
...
+    mkdir -p "${registry_config_dir}"
+    {
+      # NOTE: we need literal double quotes around some of these values
+      echo 'server="'"${KUBERNETES_REGISTRY_PULL_THROUGH_HOST}"'"'
+      echo ''
+      echo '[host."'"${KUBERNETES_REGISTRY_PULL_THROUGH_HOST}"'"]'
+      echo '  override_path = true'
+      echo '  capabilities = ["pull", "resolve"]'
+      # TODO: this is a hack. https://github.com/containerd/containerd/issues/7385
+      echo '[host."'"${KUBERNETES_REGISTRY_PULL_THROUGH_HOST}"'".headers]'
```
Comment (BenTheElder) [inline cluster/gce/gci/configure-helper.sh:3337]: /hold
this should be .header

**Comment by BenTheElder** [comment]
fixed the headers/header bug, but I also want to setup having a small test presubmit actually use this, will go get that ready in test-infra / k8s infra

**Comment by dims** [comment]
/priority critical-urgent
/triage accepted
/milestone v1.31

cc @kubernetes/release-managers 

Folks, we need to land this to unblock the 5K tests. fyi. Please don't remove hold yet until Ben is happy with additional testing.

File: cluster/gce/gci/configure-helper.sh [hunk 6]
```diff
@@ -3315,16 +3315,32 @@ oom_score = -999
...
-[plugins."io.containerd.grpc.v1.cri".registry.mirrors."k8s.gcr.io"]
-  endpoint = ["https://registry.k8s.io", "https://k8s.gcr.io",]
 [plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runc.options]
   SystemdCgroup = ${systemdCgroup}
+# enable hosts config
+[plugins."io.containerd.grpc.v1.cri".registry]
+  config_path = "/etc/containerd/certs.d"
 EOF
 
+  # used for 5k node scale tests with private pull-through cache
```
Comment (dims) [inline cluster/gce/gci/configure-helper.sh:3325]: tested by hand, looks good 

image

**Comment by BenTheElder** [comment]
https://github.com/kubernetes/test-infra/pull/33158 will allow manually vetting this fully

**Comment by BenTheElder** [comment]
running in https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/126448/pull-kubernetes-e2e-gce-pull-through-cache/1818421695510220800

**Comment by BenTheElder** [comment]
This is working as intended now. I can see the images being pulled through the cache in the test job and peeked into the e2e cluster nodes to confirm the generated config is as intended.

File: cluster/gce/util.sh [hunk 7]
```diff
@@ -1254,7 +1254,13 @@ AUTH_PROVIDER_GCP_STORAGE_PATH: $(yaml-quote "${AUTH_PROVIDER_GCP_STORAGE_PATH}"
 AUTH_PROVIDER_GCP_VERSION: $(yaml-quote "${AUTH_PROVIDER_GCP_VERSION}")
 AUTH_PROVIDER_GCP_LINUX_BIN_DIR: $(yaml-quote "${AUTH_PROVIDER_GCP_LINUX_BIN_DIR}")
 AUTH_PROVIDER_GCP_LINUX_CONF_FILE: $(yaml-quote "${AUTH_PROVIDER_GCP_LINUX_CONF_FILE}")
+KUBERNETES_REGISTRY_PULL_THROUGH_HOST: $(yaml-quote "${KUBERNETES_REGISTRY_PULL_THROUGH_HOST:-}")
 EOF
+  if [[ -n "${KUBERNETES_REGISTRY_PULL_THROUGH_BASIC_AUTH_TOKEN_PATH:-}" ]]; then
+    cat >>"$file" <<EOF
+KUBERNETES_REGISTRY_PULL_THROUGH_BASIC_AUTH_TOKEN: $(yaml-quote "$(cat "${KUBERNETES_REGISTRY_PULL_THROUGH_BASIC_AUTH_TOKEN_PATH}")")
```
Comment (BenTheElder) [inline cluster/gce/util.sh:1261]: We have to plumb this through to the nodes by placing it in the big env file we copy up here.

It does not appear to be stored in uploaded ARTIFACTS or logged in logs.

**Comment by BenTheElder** [comment]
And now https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/126448/pull-kubernetes-e2e-gce-pull-through-cache/1818444296806731776 has passed with these enabled, and I can see many images populated into the cache:
<details>
Screenshot 2024-07-30 at 6 22 00 PM
</details>

This should let us isolate high demand from scale testing in https://github.com/kubernetes/kubernetes/issues/126366

We would want end-users to do something similar, and we have docs for this in https://registry.k8s.io

**Comment by dims** [comment]
/approve
/lgtm

thanks @BenTheElder

**Pull request merged**